direct.mk

```diff
@@ -144,7 +144,10 @@ dep-validate:
 go.mod:
 	@printf "module github.com/docker/swarmkit\n\ngo 1.17\n" > $@
 
+GOGO_PROTOBUF_PATH=$(shell go env GOPATH)/pkg/mod/github.com/gogo/protobuf@$(shell go list -modfile vendor.mod -m -f '{{.Version}}' github.com/gogo/protobuf)/protobuf
+
 .PHONY: go-mod-vendor
 go-mod-vendor: go.mod
-	@GO111MODULE=on go mod vendor -modfile vendor.mod
-
+	@go mod vendor -modfile vendor.mod
+	@# ensure that key protobuf spec files are in vendor dir
```

Comment:
You mentioned these files were _intentionally_ not vendored by `go mod vendor` - out of curiosity; is that intentionally by the google/protobuf maintainers, or `go mod` ?

Reply:
The Go team believes that files that don't belong to a Go package should not belong in vendor...

Reply:
Right, so like `.appveyor.yml` and `.circleci.yml` and random `.json`,`.yaml` and Makefiles, because they remove those .....right? 😉 

I guess it's their way of telling "we don't like vendor, so we break things if you use vendor, but not if you use module cache" 🤷‍♂️

Reply:
In my understanding those random files are kept as they are contained in top-level dirs, which do contain Go code...

Reply:
>  so we break things if you use vendor

The thing is that different vendoring tools offered different behaviour, some had prune exclusion/inclusion parameters with defaults etc, so Go team chose to basically avoid having parameters... I think a sensible thing to do will be to add parameters, cause this is clearly messy as is.

Reply:
> they are contained in top-level dirs, which do contain Go code...

Ah, right. If there's no `.go` files at all in there, then it makes somewhat sense (it's odd that they remove README.md files in some cases though; I recall `vndr` explicitly kept those as they sometimes mention copyright etc).

So, one solution for such cases would be to add a `doc.go`, but it likely still would need to be imported somewhere to prevent being pruned.

Reply:
> So, one solution for such cases would be to add a `doc.go`, but it likely still would need to be imported somewhere to prevent being pruned.

I don't think it prunes unimproved packages, but cannot say for sure...